hibernate-core/src/test/java/org/hibernate/orm/test/query/hql/FunctionTests.java

```diff
@@ -464,7 +480,7 @@ public void testStrFunction(SessionFactoryScope scope) {
 							.list();
 					assertThat( session.createQuery("select str(69)").getSingleResult(), is("69") );
 					assertThat( session.createQuery("select str(date 1911-10-09)").getSingleResult(), is("1911-10-09") );
-					assertThat( session.createQuery("select str(time 12:13:14)").getSingleResult(), is("12:13:14") );
+					assertThat( session.createQuery("select str(time 12:13:14)").getSingleResult(), anyOf( is( "12:13:14"), is( "12:13:14.0000") ) );
```

Comment:
I'd rather fix the string formatting of firebird to have consistent semantics of the functions.

Reply:
I could also change it to do a similar check as on line 426:
```java
assertThat( (String) session.createQuery("select cast(datetime 1911-10-09 12:13:14 as String)").getSingleResult(), startsWith("1911-10-09 12:13:14") );
```

Reply:
I didn't check the details, but if the problem is, that we are using a cast function or something that produces the precision, and there is no alternative for Firebird, then we should use `substring` to achieve a consistent output.

Reply:
Rrrrm I dunno about that. It really might be pretty hard to write a robust date parser in SQL.

I'm not sure I think this is a big deal.

Reply:
The problem is that Firebird's `TIME` type has a fixed precision of 4 (100 micro-seconds) and conversion to string will always return the full precision. It looks like a lot of other database systems either have a fixed precision of 0, or at least a default precision of 0, or maybe do not include trailing zeroes. 

I think that removing the fractional seconds with substring or similar would not be a good solution, as it could lead to loss of information.

Reply:
I’m finding it hard to care about this one.

Reply:
AFAIU `java.sql.Time` which is what is backing the TIME type is defined to have second precision, so IMO the cast pattern from `CastType.TIME` to `CastType.STRING` should be `substring(cast(?1 as ?2) from 1 for 8)` and the test changes should be reverted.

Reply:
If someone wants a `java.sql.Time` (note though that `java.time.LocalTime` is now the recommended type), then you wouldn't cast here. The intent is to obtain a string representation of the time value, and in the case of Firebird, that string representation includes the fractional seconds.

Reply:
>The problem is that Firebird's TIME type has a fixed precision of 4 (100 micro-seconds) and conversion to string will always return the full precision. It looks like a lot of other database systems either have a fixed precision of 0, or at least a default precision of 0, or maybe do not include trailing zeroes.

I think you are right and the current behavior is ok. The problem really seems to be that other databases allow specifying the precision or use a default precision of 0. If other databases would use a different precision, we'd probably get the same output.

As far as I can say, the SQL standard defines that the time type should allow specifying a precision and default to 0. So maybe we need to adapt the Dialects in that regard to allow specifying a precision. What do you think @gavinking? Here is a section from the SQL Standard from 92 section 6.1:

>         <datetime type> ::=
>                DATE
>              | TIME [ <left paren> <time precision> <right paren> ]
>              [ WITH TIME ZONE ]
>              | TIMESTAMP [ <left paren> <timestamp precision> <right paren> ]
>              [ WITH TIME ZONE ]

>         25)If <time precision> is not specified, then 0 is implicit. If
>            <timestamp precision> is not specified, then 6 is implicit.

@mrotteveel It seems to me that using a precision of 3 as default for time is a violation of the specification.

I checked most databases and observed the following:

* Most DBs support `TIME(p)`: MySQL 5.6+, MariaDB, H2, PostgreSQL, HSQL, Ingres, Teradata, SQL Server, Sybase Anywhere, Sybase IQ, Sybase ASE, MimerSQL, CockroachDB
* SAP MaxDB supports the precision but not sure if it supports `TIME(p)` or needs `TIMESTAMP(p)`
* Oracle requires `TIMESTAMP(p)` => currently we map it to `DATE` which has no precision
* DB2 requires `TIMESTAMP(p)` => currently we map it to `TIME` which has no precision
* Informix requires `DATETIME HOUR TO FRACTION(p)` => currently we map it to `datetime hour to second`
* TimesTen requires `TIMESTAMP(p)` => currently we map it to `TT_TIME` which has no precision
* HANA requires `TIMESTAMP(p)` => currently we map it to `TIME` which has no precision
* CUBRID requires `TIMESTAMP(p)` => currently we map it to `TIME` which has no precision
* Derby requires `TIMESTAMP` which has a precision of 6 => currently we map it to `TIME` which has no precision
* Cache requires `TIMESTAMP` which has a precision of 6 => currently we map it to `TIME` which has no precision
* Spanner doesn't allow specifying precision and only supports `TIMESTAMP` with default precision 9

What do you think about registering `time($p)` by default and adding `registerColumnType( Types.TIME, 0, "time" )` to retain the current behavior, but still support a precision if needed @gavinking?

Reply:
Well I thought quite a lot about this, and I decided it's *extremely* difficult to imagine a case where anyone could possibly need a *time* with millisecond precision. (Or even *second* precision, TBH.) It really just doesn't make sense as soon as you think about what a time represents.

(*Timestamps*, of course need it.)

Reply:
FTR, the SQL standard is crap on these points. For example, the notion of a time with a timezone is just broken.

Reply:
Unfortunately, Firebird has a fixed precision of 4 for `TIME` and `TIMESTAMP`, and it doesn't support `TIME(p)` (nor `TIMESTAMP(p)`), so this workaround isn't possible.

As an aside, `CURRENT_TIME` and `CURRENT_TIMESTAMP` have a maximum precision of 3 (`CURRENT_TIME` defaults to 0), contrary to the actual datatype, see http://tracker.firebirdsql.org/browse/CORE-4458), which is why it's sometimes thought of having a precision of 3.

Reply:
OK, but IIRC, I concluded that it was perfectly fine for *Hibernate* to treat it has having precision 3. I'm pretty sure I tested this and it was working fine.

Reply:
(Or precision zero in the case of `time` as opposed to `timestamp`.)

Reply:
Yes, that will work fine, except for this specific topic: the effects of casting to string, which will include 4 digits for fractional seconds. I think this is a minor concern, that we could ignore for now (that is, accept the status quo), and maybe handle it later if and when a format function (http://tracker.firebirdsql.org/browse/CORE-6507) is available.

Reply:
So are you saying that casting from `TIME` to `STRING` should always trim the fractional parts? I never really used `TIME(p)` except for stuff like e.g. delivery time which only has minute precision, so I don't know about use cases either. I can imagine though that someone might want to track elapsed time as `TIME(p)`. The alternative would be to use `INT` with some default unit and do conversions, or use a `TIMESTAMP(p)` and ignore the date part. IMO supporting `TIME(p)` is in the same ball park as supporting `FLOAT(p)` (for which I also can't imagine a use case) which you added @gavinking, so why treat this differently?

Reply:
> So are you saying that casting from `TIME` to `STRING` should always trim the fractional parts?

As I said above, I just don't really care about this.

> I can imagine though that someone might want to track elapsed time as `TIME(p)`.

They should definitely not use `time` for that! Just think about it.

> IMO supporting `TIME(p)` is in the same ball park as supporting `FLOAT(p)`

I don't think that's similar at all. We use the `p` in `float(p)` to indicate whether we're dealing with single precision or double precision floating point types.

There's no reasonable usecase that I can imagine for millisecond-precise times. It just doesn't make sense. You're saying that something happens at time 1:35:12.1234 each day!? No, you're wrong. It doesn't.

Reply:
> You're saying that something happens at time 1:35:12.1234 each day!?

Or even: it's *going to happen at* 1:35:12.1234 next Wednesday?

Still nope. It's won't.

Reply:
@beikov I'm not saying that, but if you're insisting that the produced string value should only have second precision, then trimming the fractional parts is the only possibility for Firebird.

This is different from `FLOAT(p)` though. The standard specifies that the binary precision _p_ should lead to a database-specific floating point datatype, though usually 1 <= p <= 24 will result in a 32-bit single precision, while 25 <= p <= 53 in a 64-bit double precision, and some database systems even support higher precision. For Firebird, `FLOAT` (without precision) is a 32-bit single precision (with alias `REAL`). In Firebird 3.0 and earlier, Firebird has an undocumented `FLOAT(p)` with **decimal** precision, while Firebird 4 (based on a discussion I had with Gavin a while back), introduced `FLOAT(p)` with binary precision as defined in the standard.

What a lot of people consider `FLOAT` is actually `FLOAT(p)` with an implementation-specific default precision, which in some database systems could very well be 53 (aka a 64-bit double precision). Where `REAL` and `DOUBLE PRECISION` have a fixed (but still database-specific) precision (though usually 32-bit single and 64-bit double), `FLOAT(p)` is intended to allow for more dynamic ranges. For example, Oracle allows values of _p_ up to 126 (for a 176-bit floating point type!).

Reply:
Right, I mean, `float` and `real` don't really seem to mean anything particularly well-defined. That's why we avoid them in H6.

Reply:
>There's no reasonable usecase that I can imagine for millisecond-precise times. It just doesn't make sense. You're saying that something happens at time 1:35:12.1234 each day!? No, you're wrong. It doesn't.

The use case I was imagining was tracking elapsed time e.g. a racing car took X <time> to complete a lap.

Reply:
> trimming the fractional parts is the only possibility for Firebird

The question is: is that actually easy to do?

Reply:
> The use case I was imagining was tracking elapsed time e.g. a racing car took X [time] to complete a lap.

But that's not a time. It's a duration.

And how on earth do you propose  to handle the case where the car starts its lap at 23:59:50?

Reply:
I mean, sure, I have no doubt that programmers abuse the `time` type for all kinds of things because they like having bugs in their programs, but I don't think we should *encourage* it.

Reply:
> The use case I was imagining was tracking elapsed time e.g. a racing car took X [time] to complete a lap.

`TIME` is not intended for tracking elapsed time. Consider what happens if the elapsed time is longer than 24 hours.

> > trimming the fractional parts is the only possibility for Firebird
> 
> The question is: is that actually easy to do?

It would require string manipulation, and as such it is 1) complicated and 2) may have undesired side-effects (e.g. what if Hibernate thinks it is the wrong type, e.g. it is actually a TIMESTAMP), or what if the value is a TIME WITH TIME ZONE. 

I'm not sure the added complications are worth it. To be honest, I think wanting to cast a TIME to a string in a query is a pretty niche requirement.

Reply:
> I don't think we need to debate whether sub-second fractions of `TIME` make sense or not, this is simply the reality we're dealing with.

Well that's not clear to me.

Previously the dialect was set up to say that Firebird has 3-digit precision, because that's what is *reliably* supports everywhere (e.g. in `current_timestamp`).

I'm still not convinced that's wrong. *Why* do we need to say that it has 4-digit precision when most likely the 4th digit holds garbage? Was there some particular thing that was broken @mrotteveel? Because if there's not, I would be inclined to leave it like it was.

(Again, I don't case about what casting to string does, that's a minor side issue IMO.)

Reply:
> I find it odd that we don't allow defining the precision for time but allow mapping e.g. a Java float with a precision although nothing other than `float(24)` makes sense.

Well it's because (a) more databases support `float(p)` than support `time(p)`, and (b) we're using that syntax to distinguish single precision from double precision and it's no more work to support it generically.

On the other hand we have _no good reason_ to support `time(p)`. We're always happy with the default precision of 0.

But sure, if somebody wanted to go through all the `Dialects and add support for `time(p)` where supported then I guess I wouldn't object. I just see it as a pretty pointless exercise.

Reply:
> Well that's not clear to me.
> 
> Previously the dialect was set up to say that Firebird has 3-digit precision, because that's what it _reliably_ supports everywhere (e.g. in `current_timestamp`).

Specifically what change are you referring to? Are you referring to the change I made to `getFractionalSecondPrecisionInNanos()` from 1_000_000 to 100_000, or `getDefaultTimestampPrecision()` to return 4 (which previous to my test would apply the default value of 6)?

Reverting those won't change the reason why I made the change to the test we're commenting on. I modified that test because otherwise this specific test will fail on Firebird, no matter what precision Hibernate is attempting to apply, as cast from `TIME` to a string type will always include fractional seconds in how Firebird performs that conversion.
 
> I'm still not convinced that's wrong. _Why_ do we need to say that it has 4-digit precision when most likely the 4th digit holds garbage? Was there some particular thing that was broken @mrotteveel? Because if there's not, I would be inclined to leave it like it was.

The fourth digit will either be zero (e.g. when `CURRENT_TIME` is used), or the value that was explicitly stored there. It won't be garbage.

I have the feeling we're discussing two different things here.

Reply:
> Specifically what change are you referring to? 

Yeah, sorry, apologies, I was doing five things at once as usual and said the wrong thing. What I'm arguing is we should treat it as having consistently millisecond precision. (I might be wrong.)

> Reverting those won't change the reason why I made the change to the test we're commenting on. 

Yes, sure, I know that.

Reply:
I changed the dialect to report as if it has millisecond precision.